### ISSUE TITLE
[PSET] Method for adding a non-confidential issuance to a transaction

### ISSUE THREAD

**Issue by altafan** [vulpemventures/go-elements#94]
We need to add a new `AddIssuance` method ~~to the `*Transaction` type of the *transaction* package.~~
EDIT-2: to the `*Updater` type of the *pset* package.

This new function will have the following signature:
```
func (u *Updater) AddIssuance(arg AddIssuanceArg) (error)
```

It expects a single argument defined like:

```
type AddIssuanceArg struct {
   Precision uint,
   Contract *IssuanceContract,
   AssetAmount float64,
   TokenAmount float64,
   AssetAddress string,
   TokenAddress string,
   TokenFlag uint,
}
```

About the fields of this type, it is important to note that:
* the asset and token amount are expressed in *BTC* rather than *Sat*. After instantiating a new issuance object, the amounts in satoshis can be taken from the latter.
* the function using these args will add some outputs to the transaction (as better explained below), so it will need to get the correct scripts from the provided addresses.

Following there are the steps that the new function have to do to correctly add an unblinded issuance to a transaction that **must** contain at least 1 input:

1) Check that transaction contains at least one input
2) Instantiate a new *TxIssuanceExtended.
3)
   a)  Check that AssetAmount and AssetAddress are defined
   b) Check that TokenAmount and TokenAddress are defined if TokenAmount is > 0
4) Get the index of the first input of the transaction that has empty issuance
5) Generate entropy, asset (and token if defined) using the selected input hash and index and the provided Contract
6) Add the issuance to the designed input
7)
   a) Add an output to the transaction with asset amount, asset hash and script derived from asset address
   b) Optionally add an output with token amount, token hash and script derived from token address if the provided TokenAmount is > 0
~~8) Finally, return the index of the selected input containing the issuance, or an error if any occurs~~

EDIT-1:
I added a new field `TokenFlag uint` to the `AddIssuanceArg` type. This will be used when generating the reissuance token hash, and its values can be `0` (non confidential issuance), or `1` (confidential issuance).
In step *7b* we should also enforce that this flag has a valid value, and we should declare 2 constants `NonConfidentialReissuanceTokenFlag` and `ConfidentialReissuanceTokenFlag` for the valid values so the user can use them.

EDIT-2:
The right place for this function is the **pset** package, specifically, the updater role should be responsible for "adding an issuance" to a transaction, as simple as it may be said in this way.
The steps remain the same except for the fact that we don't need to return the issuance input index anymore, thus we return an error if occurring.
This now depends on #100, in particular, we need to able to add new outputs to the partial transaction.

**Comment by altafan** [comment]
1. You can retrieve the asset hash from the instance of `*TxIssuanceExtended`, through the method `GenerateAsset`.
2. Correct, but consider that the asset amount expected by the `AddIssuance` function is expressed in BTC, while the `AssetAmount` filed of the instance of `*TxIssuanceExtended` is instead in the confidential value format (`[9]byte{}`) and ready to be put in a tx output.
3. Yes, consider that the btcd's method you mentioned, unfortunately, is not reusable because it does not support addresses with prefixes different from the standard Bitcoin ones.

**Comment by sekulicd** [comment]
> 1. You can retrieve the asset hash from the instance of `*TxIssuanceExtended`, through the method `GenerateAsset`.
> 2. Correct, but consider that the asset amount expected by the `AddIssuance` function is expressed in BTC, while the `AssetAmount` filed of the instance of `*TxIssuanceExtended` is instead in the confidential value format (`[9]byte{}`) and ready to be put in a tx output.
> 3. Yes, consider that the btcd's method you mentioned, unfortunately, is not reusable because it does not support addresses with prefixes different from the standard Bitcoin ones.

@altafan @tiero shall we create util method PayToAddrScript in address.go?

**Comment by altafan** [comment]
>shall we create util method PayToAddrScript in address.go?

I would call this new function `ToOutputScript`, so when it's used it looks like `address.ToOutputScript`, that seems pretty readable IMO.

> Maybe we also need DecodeAddress, similar to btcutil.DecodeAddress to determine type of address and than based on type to generate script?

Correct

**Comment by tiero** [comment]
> > shall we create util method PayToAddrScript in address.go?
> 
> I would call this new function `ToOutputScript`, so when it's used it looks like `address.ToOutputScript`, that seems pretty readable IMO.
> 
> > Maybe we also need DecodeAddress, similar to btcutil.DecodeAddress to determine type of address and than based on type to generate script?
> 
> Correct

@sekulicd @altafan ACK

**Comment by altafan** [comment]
Actually, we made this function responsible for too many things, and making it working with the pset package as it is now will be quite complicated.

We should lighten this `AddIssuance` function so that by changing the 7th and 8th steps defined above.
Instead of creating and adding the outputs to the transaction, we should just create and return them along with the index of the input.
Also, the type of the returned input index should be changed to `int` so that we can return a `-1` value when an error occurs.

To summarize, the returning types of this function should change like:
```
AddIssuance(...) (int, *TxOutput, *TxOutput, error) {...}
```

Note that the second returned output should be `nil` if the provided `TokenAmount`  is `0`.
